### ISSUE TITLE
Is there a way to make pdf.save non blocking on web?

### ISSUE THREAD

**Issue by c-seeger** [DavBfr/dart_pdf#1528]
I'm currently building a large pdf with this library but when saving and downloading, it blocks the main thread. 

On mobile I can use Isolates but this will not work on web. Any suggestions here or someone that solved this already?

here is some example code:

```
import 'package:pdf/pdf.dart';
import 'package:printing/printing.dart';
import 'pdfdata.dart';
...

Future<void> generatePDF(
    BuildContext context, List<PDFData> dataList) async {
  final pw.Document pdf = pw.Document();
  try {
    for (final PDFData data in dataList) {
      await generatePDFPages(context, data, pdf);
    }
    await Printing.sharePdf(bytes: await pdf.save(), filename: 'qr-code.pdf');
  } catch (e) {
    throw Exception('pdf generation error');
  }
}

Future<void> generatedPDFPages(BuildContext context, PDFData data) async {
  pdf.addPage(....)
}
```

**Comment by c-seeger** [comment]
@DavBfr any ideas?

**Comment by DavBfr** [comment]
There is no isolates on web, so I have no solution. This code is 100% Dart so it runs on the main thread.

**Comment by bambinoua** [comment]
> @DavBfr any ideas?

Try to use [compute](https://api.flutter.dev/flutter/foundation/compute.html).

**Comment by c-seeger** [comment]
yeah there is a still [open issue](https://github.com/flutter/flutter/issues/33577) for this. If i find some time I will try to use web workers + dart2js to simulate isolation on web.

**Comment by c-seeger** [comment]
maybe [WASM](https://docs.flutter.dev/platform-integration/web/wasm) support could also work (at least for newer browsers)

**Comment by DavBfr** [comment]
Or maybe https://pub.dev/packages/isolated_worker

**Comment by c-seeger** [comment]
The main issue here is not the isolation itself since web worker can do this. The critical part is the pdf library is written in dart but web worker needs js/wasm code to run so we need to do either dart2js so web worker can use the pdf library or dart2wasm to run wasm code in web worker instead. Will check once I find some time for this issue.

**Comment by savs90** [comment]
Is it possible to create a stream to which to subscribe that returns the pdf data the moment you manipulate the pdf  (add page, text images, etc...) ?

**Comment by jpohhhh** [comment]
I can't recommend Squadron highly enough, and it should be ~perfect for this --

only issue you can run into is if you import Flutter UI code, and thus can't dart2js it, but as we see above on Nov 20th, code is 100% Dart.

(I'm actually here because syncfusion_pdf touches Flutter UI code so I can't dart2js it, and I'm trolling through the repo to triple check if there's PDF -> text parsing code.  Don't use the attached examples _directly_ because they use SyncFusion, but it's enough for you to see what I've arrived at for best practice.)

Squadron: https://pub.dev/packages/squadron
[generate_js_script.txt](https://github.com/DavBfr/dart_pdf/files/13834239/generate_js_script.txt)
[pdf_parser_service.dart.txt](https://github.com/DavBfr/dart_pdf/files/13834240/pdf_parser_service.dart.txt)

**Comment by james-majormap** [comment]
How about add pagePostProcessDoneCallback option to document.save()? ('package:pdf/widgets.dart')

It still block main thread, but has chance to make UI feedback (e.g. page processed progress)

I have extended Document class, and it seems below.

`@override
  Future<Uint8List> save({Function? pagePostProcessDoneCallback}) async {
    if (!_paint) {
      for (final page in _pages) {
        page.postProcess(this);
        if (pagePostProcessDoneCallback != null) {
          await Future.delayed(const Duration(milliseconds: 5));
          pagePostProcessDoneCallback.call();
        }
      }
      _paint = true;
    }
    return await document.save();
  }`

**Comment by c-seeger** [comment]
For reference, I was not able to fully resolve this issue (but I also did not find the time to try out squadron) but at least reduce it heavily.

The following hint gave me an idea:
https://github.com/DavBfr/dart_pdf/issues/704#issuecomment-851477894

Instead of using png or webp images I converted them to jpeg which are not additionally processed. The `pdf.save()` still eats a short amount of main thread and blocks the ui but instead of doing this for over a minute its only 1 to a few seconds.

This way I did not have to change much to the code at all, only the images used.
